### PR TITLE
fix: Add targeting key

### LINKED ISSUE #230: [BUG] Add `targeting key` to evaluation context
We need a `targeting key` in the evaluation context. It should be added to the builder, and should be immutable (like all the props in the current implementation).

---

> The evaluation context also lacks the required 'targeting key` field
>
> https://github.com/open-feature/dotnet-sdk/blob/2c237df6e0ad278ddd8a51add202b797bf81374e/test/OpenFeature.Tests/OpenFeatureEvaluationContextTests.cs#L46

_Originally posted by @roelofb in https://github.com/open-feature/dotnet-sdk/issues/204#issuecomment-1925910926_

### PR DESCRIPTION
**Pull request by roelofb**
Fixes #230

### CODE REVIEW

File: src/OpenFeature/Model/EvaluationContextBuilder.cs [hunk 1]
```diff
@@ -125,6 +138,23 @@ public EvaluationContextBuilder Set(string key, DateTime value)
         /// <returns>This builder</returns>
         public EvaluationContextBuilder Merge(EvaluationContext context)
         {
+            string newTargetingKey = "";
+
+            if (TargetingKey != null && TargetingKey.Trim() != string.Empty)
```
Comment (austindrenski) [inline src/OpenFeature/Model/EvaluationContextBuilder.cs:143]:
```suggestion
            if (!string.IsNullOrWhitespace(TargetingKey))
```

File: src/OpenFeature/Model/EvaluationContextBuilder.cs [hunk 2]
```diff
@@ -125,6 +138,23 @@ public EvaluationContextBuilder Set(string key, DateTime value)
...
+            {
+                newTargetingKey = TargetingKey;
+            }
+
+            if (context.TargetingKey != null && context.TargetingKey.Trim() != string.Empty)
```
Comment (austindrenski) [inline src/OpenFeature/Model/EvaluationContextBuilder.cs:148]:
```suggestion
            if (!string.IsNullOrWhitespace(context.TargetingKey))
```

File: src/OpenFeature/Model/EvaluationContextBuilder.cs [hunk 3]
```diff
@@ -125,6 +138,23 @@ public EvaluationContextBuilder Set(string key, DateTime value)
...
+            {
+                newTargetingKey = context.TargetingKey;
+            }
+
+            if (newTargetingKey != null && newTargetingKey.Trim() != string.Empty)
```
Comment (austindrenski) [inline src/OpenFeature/Model/EvaluationContextBuilder.cs:153]:
```suggestion
            if (!string.IsNullOrWhitespace(newTargetingKey))
```

File: src/OpenFeature/Model/EvaluationContextBuilder.cs (hunk 1 above)
Comment (toddbaert) [reply src/OpenFeature/Model/EvaluationContextBuilder.cs:143]: `IsNullOrWhitespace` is the best improvement .NET made on Java, change my mind.

File: src/OpenFeature/Model/EvaluationContextBuilder.cs [hunk 4]
```diff
@@ -125,6 +138,23 @@ public EvaluationContextBuilder Set(string key, DateTime value)
         /// <returns>This builder</returns>
         public EvaluationContextBuilder Merge(EvaluationContext context)
         {
+            string newTargetingKey = "";
+
+            if (!string.IsNullOrWhitespace(TargetingKey))
```
Comment (austindrenski) [inline src/OpenFeature/Model/EvaluationContextBuilder.cs:143]: Ugh, this is my fault, sorry :[

```suggestion
            if (!string.IsNullOrWhiteSpace(TargetingKey))
```

File: src/OpenFeature/Model/EvaluationContextBuilder.cs [hunk 5]
```diff
@@ -125,6 +138,23 @@ public EvaluationContextBuilder Set(string key, DateTime value)
...
+            {
+                newTargetingKey = TargetingKey;
+            }
+
+            if (!string.IsNullOrWhitespace(context.TargetingKey))
```
Comment (austindrenski) [inline src/OpenFeature/Model/EvaluationContextBuilder.cs:148]:
```suggestion
            if (!string.IsNullOrWhiteSpace(context.TargetingKey))
```

File: src/OpenFeature/Model/EvaluationContextBuilder.cs [hunk 6]
```diff
@@ -125,6 +138,23 @@ public EvaluationContextBuilder Set(string key, DateTime value)
...
+            {
+                newTargetingKey = context.TargetingKey;
+            }
+
+            if (!string.IsNullOrWhitespace(newTargetingKey))
```
Comment (austindrenski) [inline src/OpenFeature/Model/EvaluationContextBuilder.cs:153]:
```suggestion
            if (!string.IsNullOrWhiteSpace(newTargetingKey))
```

**Review by austindrenski — CHANGES_REQUESTED** [review]

File: src/OpenFeature/Model/EvaluationContextBuilder.cs (hunk 4 above)
Comment (austindrenski) [inline src/OpenFeature/Model/EvaluationContextBuilder.cs:143]:
```suggestion
            if (!string.IsNullOrWhiteSpace(TargetingKey))
```

File: src/OpenFeature/Model/EvaluationContextBuilder.cs (hunk 5 above)
Comment (austindrenski) [inline src/OpenFeature/Model/EvaluationContextBuilder.cs:148]:
```suggestion
            if (!string.IsNullOrWhiteSpace(context.TargetingKey))
```

File: src/OpenFeature/Model/EvaluationContextBuilder.cs (hunk 6 above)
Comment (austindrenski) [inline src/OpenFeature/Model/EvaluationContextBuilder.cs:153]:
```suggestion
            if (!string.IsNullOrWhiteSpace(newTargetingKey))
```

File: src/OpenFeature/Model/EvaluationContextBuilder.cs (hunk 4 above)
Comment (austindrenski) [inline src/OpenFeature/Model/EvaluationContextBuilder.cs:143]:
```suggestion
            if (!string.IsNullOrWhiteSpace(TargetingKey))
```

**Review by austindrenski — APPROVED** [review]

**Comment by roelofb** [comment]
Not sure how to fix this test failure, the test harness looks incredibly involved TBH

**Comment by toddbaert** [comment]
> Not sure how to fix this test failure, the test harness looks incredibly involved TBH

Ya. We have an issue to fix this. I think this is similar to https://github.com/open-feature/dotnet-sdk/pull/223#issuecomment-1927668079. I will try to fix this today.

File: test/OpenFeature.E2ETests/Steps/EvaluationStepDefinitions.cs [hunk 7]
```diff
@@ -200,12 +200,11 @@ public void Giventhevariantshouldbeandthereasonshouldbe(string expectedVariant,
...
-            attributes.Add(field1, new Value(value1));
-            attributes.Add(field2, new Value(value2));
-            attributes.Add(field3, new Value(value3));
-            attributes.Add(field4, new Value(bool.Parse(value4)));
-            this.context = new EvaluationContext(new Structure(attributes));
+            this.context = new EvaluationContextBuilder()
+                .Set(field1, value1)
+                .Set(field2, value2)
+                .Set(field3, value3)
+                .Set(field4, value4).Build();
```
Comment (toddbaert) [inline test/OpenFeature.E2ETests/Steps/EvaluationStepDefinitions.cs:207]:
```suggestion
                .Set(field4, bool.Parse(value4)).Build();
```

Comment (toddbaert) [reply test/OpenFeature.E2ETests/Steps/EvaluationStepDefinitions.cs:207]: I think this is the cause of the test failures.

**Review by toddbaert — APPROVED** [review]

**Comment by toddbaert** [comment]
I'll merge this Monday unless anyone objects!

Thanks a lot @roelofb !

**Review by benjiro — APPROVED** [review]

**Pull request merged**